proxy_map.go

```diff
@@ -0,0 +1,1067 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package hazelcast
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/hazelcast/hazelcast-go-client/internal"
+	"github.com/hazelcast/hazelcast-go-client/internal/cb"
+	"github.com/hazelcast/hazelcast-go-client/internal/event"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto/codec"
+	iproxy "github.com/hazelcast/hazelcast-go-client/internal/proxy"
+	"github.com/hazelcast/hazelcast-go-client/predicate"
+	pubserialization "github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/types"
+)
+
+const (
+	maxIndexAttributes = 255
+)
+
+type MapEntryListenerConfig struct {
+	Predicate          predicate.Predicate
+	IncludeValue       bool
+	Key                interface{}
+	NotifyEntryAdded   bool
+	NotifyEntryRemoved bool
+	NotifyEntryUpdated bool
+}
+
+type Map struct {
+	*proxy
+	refIDGenerator *iproxy.ReferenceIDGenerator
+	ctx            context.Context
+	lockID         int64
+}
+
+func newMap(ctx context.Context, p *proxy) *Map {
+	lockID := ctx.Value(lockIDKey).(int64)
+	return &Map{
+		proxy:          p,
+		refIDGenerator: iproxy.NewReferenceIDGenerator(),
+		ctx:            ctx,
+		lockID:         lockID,
+	}
+}
+
+func (m *Map) withContext(ctx context.Context) *Map {
+	lockID := int64(reflect.ValueOf(ctx).Pointer())
```

Comment:
nit: this method seems to be unused. Also, it seems weird and potentially unsafe to use context pointer's address as the lock identity. While Go has a non-moving GC, new objects may be allocated at the same address as the old, collected objects.